### PR TITLE
adjusted training service and test cases

### PR DESCRIPTION
**Pull request by ncrimi**
implemented error convergence and global validation test set.

### CODE REVIEW

File: training-service/test/hack3rz_test.py [hunk 1]
```diff
@@ -90,8 +90,9 @@ def load_test_X_T(self, lang_name):
             training data array X and target data array T. For more informationn check information in src/services/training.py
         
         """
-        X, T = data_preprocessing(self.annotation_repository.find_data_to_train_with(lang_name))
-        return X, T
+        training_data = self.annotation_repository.find_data_to_train_with(lang_name)
```
Comment (pemmenegger) [inline training-service/test/hack3rz_test.py:93]: Mmmmh, since we removed the data_preprocessing() in this function it is exactly the same than a "normal" database call. It only made sense previously because of the two test, i.e. fetch data from database and preprocess it. I think we don't need the whole function anymore because we simply can execute ```self.annotation_repository.find_data_to_train_with(lang_name)``` in our tests as it is already done at the moment. 

TODO:
- Remove this function and replace its calls in the tests with ```self.annotation_repository.find_data_to_train_with(lang_name)```directly

File: training-service/src/services/training.py [hunk 2]
```diff
@@ -1,3 +1,4 @@
+from calendar import TUESDAY
```
Comment (pemmenegger) [inline training-service/src/services/training.py:1]: Why do we need this?

File: training-service/test/test_training_service.py [hunk 3]
```diff
@@ -1,8 +1,9 @@
 import json
-from unittest.mock import patch
+from tkinter import E
+from unittest.mock import patch, MagicMock
```
Comment (pemmenegger) [inline training-service/test/test_training_service.py:3]: Why do we need MagicMock?

File: training-service/test/test_training_service.py [hunk 4]
```diff
@@ -1,8 +1,9 @@
 import json
-from unittest.mock import patch
+from tkinter import E
```
Comment (pemmenegger) [inline training-service/test/test_training_service.py:2]: Why do we need this?

File: training-service/src/services/training.py [hunk 5]
```diff
@@ -120,15 +129,25 @@ def train(model, X_train, T_train, epochs=10):
     model.setup_for_finetuning()
     X_train, T_train = shuffle_data(X_train, T_train)
     losses = np.array([])
+    delta = 100
     for epoch in range(epochs):
-        print(f'Loading {(epoch+1/epochs)*100}%', flush=True)
+        print("Epoch: ", epoch)
```
[the remark this answers is not in the document]
Comment (ncrimi) [reply training-service/src/services/training.py:134]: done

File: training-service/src/services/training.py [hunk 6]
```diff
@@ -120,15 +129,25 @@ def train(model, X_train, T_train, epochs=10):
...
+                delta = epoch_losses[idx-1]-loss_of_sample
+                print("current delta: ", delta, flush=True)
             epoch_losses = np.append(epoch_losses, loss_of_sample)
-        avg_epoch_loss = np.mean(epoch_losses)
+        if len(epoch_losses) !=0:
+            avg_epoch_loss = np.mean(epoch_losses)
+            print(f'Average Loss {avg_epoch_loss} in epoch {epoch+1}', flush=True)  
         losses = np.append(losses, avg_epoch_loss)
-        print(f'Average Loss {avg_epoch_loss} in epoch {epoch+1}', flush=True)
+        if delta < error_convergence:
```
[the remark this answers is not in the document]
Comment (ncrimi) [reply training-service/src/services/training.py:148]: `if len(epoch_losses) !=0:`is used to prevent calculating the mean on an empty array (epoch_losses). But you are right. If we switch the if statements we should not run into this problem.

File: training-service/requirements. [hunk 7]
```diff
@@ -0,0 +1,48 @@
+apispec==5.1.1
```
Comment (pemmenegger) [inline training-service/requirements.:1]: wrong requirements.txt export

1. we only have a requirements.txt in the root folder, please delete the requirements. in the test folder
2. requirements. != requirements.txt (wrong file format export?), command for the export should be ```pip freeze > requirements.txt``` from the training-service directory

Comment (ncrimi) [reply training-service/requirements.:1]: okay got it

**Review by pemmenegger — APPROVED** [review]
